packages/suite/src/views/wallet/transactions/TransactionList/useFetchTransactions.ts

```diff
@@ -0,0 +1,223 @@
+                await fetchCommon(pagesFetched + 1, {
+                    recursive: true,
+                });
+                setFetching(false);
+                setFetchedAll(true);
+            }),
+        [synchronize, fetchCommon, pagesFetched, fetchedAll],
+    );
+
+    return { fetchNext, pagesFetched, fetchPage, fetchAll, isFetching, fetchedAll };
+};
+
+export const useVisibleTransactionsRetriever = ({
+    account,
+    numberOfPagesRequested,
+}: {
+    account: Account;
+    numberOfPagesRequested: number;
+}) => {
+    const allTransactions = useSelector(state =>
+        selectAccountTransactionsWithNulls(state, account.key),
+    );
+    const { discovery } = useDiscovery();
+
+    const {
+        fetchedAll,
+        isFetching,
+        pagesFetched: allTransactionsPagesFetched,
+        fetchPage,
+    } = useFetchTransactions(account, allTransactions);
+    const allAccountTransactions = useSelector(state =>
+        selectAccountTotalTransactions(state, account.key),
+    );
+    const transactionsIsLoading = useSelector(state =>
+        selectIsLoadingAccountTransactions(state, account.key),
+    );
+    const tokenDefinitions = useSelector(state =>
+        selectNetworkTokenDefinitions(state, account.symbol),
+    );
+
+    const visibleTransactions = useMemo(
+        () =>
+            tokenDefinitions
+                ? allTransactions.filter(
+                      transaction => !getIsPhishingTransaction(transaction, tokenDefinitions),
+                  )
+                : allTransactions,
+        [allTransactions, tokenDefinitions],
+    );
+
+    const perPage = getTxsPerPage(account.networkType);
+    const numberOfHidden = allTransactions.length - visibleTransactions.length;
+    const totalPossiblyVisible = allAccountTransactions - numberOfHidden;
+
+    useEffect(() => {
+        if (
+            shouldAttemptToLoadNextPageForVisibleTransactions({
+                totalNumberOfTransactions: allAccountTransactions,
+                currentNumberOfVisibleTransactions: visibleTransactions.length,
+                currentNumberOfTransactions: allTransactions.length,
+                perPage,
+                numberOfPagesRequested,
+            })
+        ) {
+            fetchPage(allTransactionsPagesFetched + 1);
+        }
+    }, [
+        account.networkType,
+        allAccountTransactions,
+        allTransactions.length,
+        allTransactionsPagesFetched,
+        fetchPage,
+        fetchedAll,
+        numberOfPagesRequested,
+        perPage,
+        totalPossiblyVisible,
+        visibleTransactions.length,
+    ]);
+
+    return {
+        allTransactions,
+        visibleTransactions,
+        visibleTotal: totalPossiblyVisible,
+        isFetching: isFetching || transactionsIsLoading,
+        isLoading: discovery && discovery?.status !== DiscoveryStatus.COMPLETED, // NOTE: loading indicates that the state of the data is unknown and we are "loading for the first time"
```

Comment:
I am not entirely sure about this.

`isLoading` should indicate a state, when the wallet is loading the transactions for the first time. This is useful when displaying the empty state.